bluetooth:host: added a mode param to PAST parameters

roadrunner-97 opened the pull request:
We now have a parameter for the mode to specify if we sync or not, whether to generate sync reports, and whether those reports should be duplicate filtered. also adds to the validation for this mode param.

Signed-off-by: Raphael Treccani-Chinelli <raphael.treccani@nordicsemi.no>

hermabe reviewed (commented):
Changes look good, however all in-tree users must also be updated to set the `mode` field. I see two samples and the shell need updating.

Because it is an API change we should also add a line in the changelog saying that users should set the `mode=BT_HCI_LE_PAST_MODE_SYNC_REPORTS_UNFILTERED ` to keep old behavior and that the mode defines have been renamed.

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1572,6 +1572,13 @@ enum {
 };
 
 struct bt_le_per_adv_sync_transfer_param {
+	/**
+	 * @brief Mode
+	 *
+	 * Determines whether or not to sync to the PAST indication, whether or not
+	 * to generate sync reports, and if the reports should be filtered or not.
+	*/
+	uint8_t mode;
```
I would suggest to add the mode to the already existing `options` field instead. 

Something along the lines of `BT_LE_PER_ADV_SYNC_TRANSFER_OPT_NO_SYNC` and `BT_LE_PER_ADV_SYNC_TRANSFER_OPT_NO_REPORTS` (and document those options to be mutually exclusive). 

That way it won't be an API change and be fully backwards compatible, as well as following the structure of not only the sync parameters, but the rest of the Bluetooth API.

rtreccani replied:
I've added them as flags, with documentation.

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -1766,9 +1766,10 @@ struct bt_hci_rp_le_per_adv_set_info_transfer {
-#define BT_HCI_LE_PAST_MODE_NO_SYNC              0x00
-#define BT_HCI_LE_PAST_MODE_NO_REPORTS           0x01
-#define BT_HCI_LE_PAST_MODE_SYNC                 0x02
+#define BT_HCI_LE_PAST_MODE_NO_SYNC                 0x00
+#define BT_HCI_LE_PAST_MODE_SYNC_NO_REPORTS         0x01
+#define BT_HCI_LE_PAST_MODE_SYNC_REPORTS_UNFILTERED 0x02
+#define BT_HCI_LE_PAST_MODE_SYNC_REPORTS_FILTERED   0x03
```
This is modifying values in a public API file. 

I would suggest to simply add `BT_HCI_LE_PAST_MODE_SYNC_FILTERED` instead of modifying `BT_HCI_LE_PAST_MODE_SYNC` as well.

rtreccani replied:
I've removed the changes, now only adding the new value (and formatting the old ones)

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1569,6 +1569,19 @@ enum {
 	/** Only sync to packets with constant tone extension */
 	BT_LE_PER_ADV_SYNC_TRANSFER_OPT_SYNC_ONLY_CTE = BIT(3),
+
+	/**
+	 * The following 2 options are mutually exclusive. You should set
+	 * at most one of these options at once. If no options are set, PAST
+	 * packets will be synced to and sync reports will be generated without
+	 * duplicate filtering.
+	*/
```
This information should be placed in the documentation of the relevant options. The way it is added here probably won't result in good doxygen documentation.

rtreccani replied:
Do you mean it should be above the definition of each bit?

Thalley replied:
Yes. They could just reference each other, stating that the other bit shall not be set at the same time :)

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1569,6 +1569,19 @@ enum {
+	 * duplicate filtering.
+	*/
+
+	/** Sync to received PAST packets but don't generate sync reports*/
```
```suggestion
	/** Sync to received PAST packets but do not generate sync reports*/
```

Thalley commented on subsys/bluetooth/host/scan.c:
```diff
@@ -1707,6 +1707,10 @@ static bool valid_past_param(
 	    param->timeout > 0x4000) {
 		return false;
 	}
+	if ((param->options & BT_LE_PER_ADV_SYNC_TRANSFER_OPT_REPORTING_INITIALLY_DISABLED) &&
+	   (param->options & BT_LE_PER_ADV_SYNC_TRANSFER_OPT_FILTER_DUPLICATES)) {
```
```suggestion
	if ((param->options & BT_LE_PER_ADV_SYNC_TRANSFER_OPT_REPORTING_INITIALLY_DISABLED) &&
	    (param->options & BT_LE_PER_ADV_SYNC_TRANSFER_OPT_FILTER_DUPLICATES)) {
```

Thalley reviewed (changes requested):
Minor requests, otherwise LGTM

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1569,6 +1569,22 @@ enum {
+	 * @brief Sync to received PAST packets but don't generate sync reports
+	 *
+	 * This option must not be set at the same time as
+	 * `BT_LE_PER_ADV_SYNC_TRANSFER_OPT_FILTER_DUPLICATES`.
```
```suggestion
	 * @ref BT_LE_PER_ADV_SYNC_TRANSFER_OPT_FILTER_DUPLICATES.
```

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1569,6 +1569,22 @@ enum {
+	 * @brief Sync to received PAST packets and generate sync reports with duplicate filtering
+	 *
+	 * This option must not be set at the same time as
+	 * `BT_LE_PER_ADV_SYNC_TRANSFER_OPT_REPORTING_INITIALLY_DISABLED`.
```
```suggestion
	 * @ref BT_LE_PER_ADV_SYNC_TRANSFER_OPT_REPORTING_INITIALLY_DISABLED.
```

Thalley reviewed (changes requested):
One final minor comment, otherwise LGTM

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1569,6 +1569,22 @@ enum {
+	 * @brief Sync to received PAST packets but don't generate sync reports
+	 *
+	 * This option must not be set at the same time as
+	 * @ref BT_LE_PER_ADV_SYNC_TRANSFER_OPT_FILTER_DUPLICATES .
```
```suggestion
	 * @ref BT_LE_PER_ADV_SYNC_TRANSFER_OPT_FILTER_DUPLICATES.
```

rtreccani replied:
I wasn't sure if doxygen would be ok with a full stop running into the variable name, that's why i left the space. I can remove  if you want :)

Thalley replied:
It works :) See e.g. https://docs.zephyrproject.org/latest/connectivity/bluetooth/api/gap.html#c.@40.BT_LE_ADV_OPT_NO_2M that has `@ref BT_LE_ADV_OPT_EXT_ADV.`

Thalley commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -1569,6 +1569,22 @@ enum {
+	 * @brief Sync to received PAST packets and generate sync reports with duplicate filtering
+	 *
+	 * This option must not be set at the same time as
+	 * @ref BT_LE_PER_ADV_SYNC_TRANSFER_OPT_REPORTING_INITIALLY_DISABLED .
```
```suggestion
	 * @ref BT_LE_PER_ADV_SYNC_TRANSFER_OPT_REPORTING_INITIALLY_DISABLED.
```

Thalley reviewed (dismissed):
Two minor nitpicking comments, but is OK unless someone else wants to fix that too

roadrunner-97 committed:
bluetooth: host: Added extra options to PAST subscribe

allow disabling reports (synchronise but don't generate sync reports)
and allows enabling sync reporting with filtering of duplicates.
the default option remains to establish sync, with sync reports,
but without duplicate filtering.

Signed-off-by: Raphael Treccani-Chinelli <raphael.treccani@nordicsemi.no>
